dart/common/Uri.cpp

```diff
@@ -0,0 +1,435 @@
+    output << "//" << *mAuthority;
+
+  output << mPath.get_value_or("");
+
+  if(mQuery)
+    output << "?" << *mQuery;
+
+  if(mFragment)
+    output << "#" << *mFragment;
+
+  return output.str();
+}
+
+bool Uri::fromStringOrPath(const std::string& _input)
+{
+  if (!fromString(_input))
+    return false;
+
+  // Assume that any URI without a scheme is a path.
+  if (!mScheme && mPath)
+  {
+    mScheme = "file";
+
+    // Replace backslashes (from Windows paths) with forward slashes.
+    std::replace(std::begin(*mPath), std::end(*mPath), '\\', '/');
+  }
+
+  return true;
+}
+
+std::string Uri::getUri(const std::string& _input)
+{
+  Uri uri;
+  if(uri.fromStringOrPath(_input))
+    return uri.toString();
+  else
+    return "";
+}
+
+std::string Uri::getRelativeUri(
+  const std::string& _base, const std::string& _relative, bool _strict)
+{
+  Uri baseUri;
+  if(!baseUri.fromString(_base))
+  {
+    dtwarn << "[getRelativeUri] Failed parsing base URI '"
+           << _base << "'.\n";
+    return "";
+  }
+
+  Uri relativeUri;
+  if(!relativeUri.fromString(_relative))
+  {
+    dtwarn << "[getRelativeUri] Failed parsing relative URI '"  
+           << _relative << "'.\n";
+    return "";
+  }
+
+  Uri mergedUri;
+  if(!mergedUri.fromRelativeUri(baseUri, relativeUri, _strict))
+  {
+    dtwarn << "[getRelativeUri] Failed merging URI '" << _relative
+           << "' with base URI '" << _base << "'.\n";
+    return "";
+  }
+
+  return mergedUri.toString();
+}
+
+std::string Uri::mergePaths(const Uri& _base, const Uri& _relative)
+{
+  assert(_base.mPath && "The path component is always defined.");
+  assert(_relative.mPath && "The path component is always defined.");
+
+  // This directly implements the logic from Section 5.2.3. of RFC 3986.
+  if(_base.mAuthority && _base.mPath->empty())
+    return "/" + *_relative.mPath;
+  else
+  {
+    const size_t index = _base.mPath->find_last_of('/');
+    if(index != std::string::npos)
+      return _base.mPath->substr(0, index + 1) + *_relative.mPath;
+    else
+      return *_relative.mPath;
+  }
+}
+
+std::string Uri::removeDotSegments(const std::string& _path)
+{
+  // 1.  The input buffer is initialized with the now-appended path
+  //     components and the output buffer is initialized to the empty
+  //     string.
+  std::string input = _path;
+  std::string output;
+
+  // 2.  While the input buffer is not empty, loop as follows:
+  while(!input.empty())
+  {
+    // A.  If the input buffer begins with a prefix of "../" or "./",
+    //     then remove that prefix from the input buffer; otherwise,
+    if(startsWith(input, "../"))
+    {
+      input = input.substr(3);
+    }
+    else if(startsWith(input, "./"))
+    {
+      input = input.substr(2);
+    }
+    // B.  if the input buffer begins with a prefix of "/./" or "/.",
+    //     where "." is a complete path segment, then replace that
+    //     prefix with "/" in the input buffer; otherwise,
+    else if(input == "/.")
```

Comment:
This first `if` statement implements the "or "/.", where "." is a complete path segment" clause. I believe this can only happen if the path ends with `/.`. In the case you described, the path segment is `.hiddenFile.txt`.

These comments are copied and pasted from [Section 5.2.4 of RFC-3986](https://tools.ietf.org/html/rfc3986#section-5.2.4), so we'd have to chat with Tim Berners-Lee if it's incorrect. :smile: